clippy_lints/src/loops/manual_find.rs

```diff
@@ -0,0 +1,189 @@
+                            snippet_with_applicability(cx, pat.span, "..", &mut applicability),
+                            snippet_with_applicability(cx, binding.span, "..", &mut applicability),
+                        )[..],
+                    );
+                }
+                let mut borrows = 0;
+                match cond.kind {
+                    ExprKind::Binary(_, l, r) => {
+                        for side in [l, r] {
+                            if !matches!(
+                                side.kind,
+                                ExprKind::MethodCall(_, [caller, ..], _) if ident_in_expr(cx, ident, caller)
+                            ) && ident_in_expr(cx, ident, side)
+                            {
+                                borrows = 1;
+                            }
+                        }
+                    },
+                    ExprKind::Call(..) => borrows = 1,
+                    ExprKind::MethodCall(_, [caller, ..], _) => {
+                        if !ident_in_expr(cx, ident, caller) {
+                            borrows = 1;
+                        }
+                    },
+                    _ => {},
+                }
+                if is_ref_to_binding {
+                    borrows += 1;
+                }
+                snippet.push_str(
+                    &format!(
+                        ".find(|{}{}| {})",
+                        "&".repeat(borrows),
+                        snippet_with_applicability(cx, binding.span, "..", &mut applicability),
+                        snippet_with_applicability(cx, cond.span, "..", &mut applicability),
+                    )[..],
+                );
+                if is_ref_to_binding {
+                    snippet.push_str(".copied()");
+                }
+                span_lint_and_sugg(
+                    cx,
+                    MANUAL_FIND,
+                    span.with_hi(ret_exprs[1].span.hi()),
+                    "manual implementation of Iterator::find",
+                    "replace with an iterator",
+                    snippet,
+                    applicability,
+                );
+            }
+        }
+    }
+}
+
+fn ident_of(expr: &Expr<'_>) -> Option<Ident> {
+    if let ExprKind::Path(QPath::Resolved(_, path)) = expr.kind {
+        if let [seg, ..] = path.segments {
+            return Some(seg.ident);
+        }
+    }
+    None
+}
+
+fn ident_in_expr<'tcx>(cx: &LateContext<'tcx>, ident: Ident, expr: &'tcx Expr<'_>) -> bool {
+    let mut contains = false;
+    expr_visitor(cx, |e| {
+        contains |= ident_of(e).map_or(false, |i| i == ident);
+        !contains
+    })
+    .visit_expr(expr);
+    contains
+}
+
+#[derive(Default)]
+struct PatIdentCollector {
+    idents: Vec<Ident>,
+}
+
+impl PatIdentCollector {
+    fn visit_pat(&mut self, pat: &Pat<'_>) {
+        match pat.kind {
+            PatKind::Binding(BindingAnnotation::Unannotated, _, ident, None) => self.idents.push(ident),
```

Comment:
If there are idents we aren't handling that can cause problems, here `.idents` would have length 1

```rust
fn mixed_binding_modes(arr: Vec<(i32, String)>) -> Option<i32> {
    for (x, mut s) in arr {
        if x == 1 && s.as_mut_str().len() == 2 {
            return Some(x);
        }
    }
    None
}
```

We could have some function that returns `Option<Ident>` implemented using [`Pat::each_binding`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Pat.html#method.each_binding), if multiple idents appear or there's  one we aren't going to handle (`mut/ref/ref mut`) then it should return `None`